### PR TITLE
PUBDEV-3281: ARFF parser parses attached file incorrectly (by Vlad)

### PR DESCRIPTION
**Pull request by vpatryshev**

### CODE REVIEW

File: h2o-algos/src/main/java/hex/pca/PCA.java [hunk 1]
```diff
@@ -315,6 +315,9 @@ public void computeImpl() {
           // Note: Singular values ordered in weakly descending order by algorithm
           _job.update(1, "Calculating SVD of Gram matrix locally");
           Matrix gramJ = _wideDataset ? new Matrix(ogtsk._gram.getXX()) : new Matrix(gtsk._gram.getXX());
+          final String msg = "Matrix dimensions: " + gramJ.getRowDimension() + "x" + gramJ.getColumnDimension();
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/pca/PCA.java]: Is this change related to PUBDEV-3281?

Why is the println here?

File: h2o-core/src/main/java/water/parser/CsvParser.java [hunk 2]
```diff
@@ -22,7 +22,19 @@
...
+  private static int skipHeader(byte[] bits, int offset) {
+    int offsetToNonspace = skipSpaces(bits, offset);
+    while (offsetToNonspace < bits.length && isCommentChar(bits[offsetToNonspace])) {
+      offset = skipToNewLine(bits, offsetToNonspace) + 1;
+      offsetToNonspace = skipSpaces(bits, offset);
+    }
+    return offset;
+  }
+  
+  private int patience = 20;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: Is this used?

File: h2o-core/src/main/java/water/parser/CsvParser.java [hunk 3]
```diff
@@ -57,25 +69,13 @@
...
-        }
-        if ((offset + 1 < bits.length) && (bits[offset] == CHAR_CR) && (bits[offset + 1] == CHAR_LF)) ++offset;
-        ++offset;
-//        System.out.println();
-        if (offset >= bits.length)
-          return dout;
-        c = bits[offset];
-      }
+    if (cidx == 0) {
+      offset = skipHeader(bits, offset);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: Is this just refactoring or was anything changed in the implementation?

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 4]
```diff
@@ -27,6 +28,8 @@
   public static final int HAS_HEADER = 1;
   public static final int GUESS_COL_CNT = -1;
 
+  public BytesStats bytesStats;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: What is the purpose of bytesStats? Is it just for the ARFF parser? If so we should probably have a derived ParseSetup for ARFF parser and put it there,  see AvroParseSetup

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 5]
```diff
@@ -267,11 +273,16 @@ public static ParseSetup guessSetup( Key[] fkeys, ParseSetup userSetup ) {
...
     if (ice instanceof Frame && ((Frame) ice).vec(0) instanceof UploadFileVec) {
       t._gblSetup._chunk_size = FileVec.DFLT_CHUNK_SIZE;
     } else {
-      t._gblSetup._chunk_size = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int calculatedChunkSize = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
+          Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int maxExpectedHeaderSize = Math.max(10000, (int)t._maxLineLength * 300); // to make sure we dont lose bits and get a negative
+      int acceptableHeaderSize = Math.min(500000, maxExpectedHeaderSize); //can't expect half a meg of headers
+      int finalChunkSize = Math.max(calculatedChunkSize, acceptableHeaderSize);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This might be a controversial change. Why is chunkSize function of a header size? Does this affect binary parsers as well?

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 6]
```diff
@@ -346,13 +357,14 @@ public GuessSetupTsk(ParseSetup userSetup) {
...
-/*        if (ice instanceof ByteVec
-                || ((Frame)ice).vecs()[0] instanceof UploadFileVec
-                || bv.length() <= FileVec.DFLT_CHUNK_SIZE
-                || decompRatio > 1.0) { */
         try {
           _gblSetup = guessSetup(bv, bits, _userSetup);
+          _gblSetup.bytesStats = new BytesStats(bits);
+          {
+            if (_maxLineLength != _gblSetup.bytesStats.maxWidth) {
+//TODO(vlad): investigate! the length produced by maxLineLength is 1.3 meg, and the line consists of zero chars            throw new IllegalStateException("wtf");
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This should probably be resolved in this PR or better explained what is happening. TODOs should be clear to anyone.

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 7]
```diff
@@ -346,13 +357,14 @@ public GuessSetupTsk(ParseSetup userSetup) {
         _maxLineLength = maxLineLength(bits);
         if (_maxLineLength==-1) throw new H2OIllegalArgumentException("The first 4MB of the data don't contain any line breaks. Cannot parse.");
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: this will never be satisfied because you changed maxLineLength to return 0 instead of -1

File: h2o-algos/src/main/java/hex/pca/PCA.java (hunk 1 above)
Comment (vpatryshev) [inline h2o-algos/src/main/java/hex/pca/PCA.java]: The stuff was failing intermittently; anyway, it's irrelevant now, will revert.

File: h2o-core/src/main/java/water/parser/CsvParser.java (hunk 2 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: Same here; will revert. Thanks.

File: h2o-core/src/main/java/water/parser/CsvParser.java (hunk 3 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/CsvParser.java]: It's refactoring, one small step to make code more manageable.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 4 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: It has a value independent from ARFF; it's like wc. And it's already in master, by the way.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 5 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: The current algorithm requires that the whole header fits in one chunk. That was the problem this change fixes. Later I'll go ahead and use a better algorithm; there are pieces of code that were probably developed for that purpose. Have to work on it.
This specific change kind of makes sure that a reasonable-sized header will fit into one chunk. Making small chunks makes no sense when reading files, it's 2017.
Still, since maxLineLength is, at times, calculated incorrectly (I have some examples), we have to be careful.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 6 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: That's exactly what the TODO says: I have an example where maxLineLength is erroneously calculated at about 1.3 meg. We can't fix all existing bugs in one PR. I'll fix it, but later on.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 7 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Right, thank you. The whole exception was misleading. -1 did not signal the file being too wide anyway. Removing it.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 4 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: I didn't find BytesStats in master. What value does it have for the other parsers?

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 5 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Yes, it is 2017 - but we have a function that is supposed to calculate the optimal chunk size FileVec.calcOptimalChunkSize(...). This function contains the heuristics that determines the chunk size. If you are overriding the result of the function you are by _definition of the function_ creating a sub-optimal chunk size (assuming the calcOptimalChunkSize function really returns the optimal chunk size, that would be another discussion).

What part of the ARFF parser requires the header to fit in just one chunk? Is it the guessSetup function? Please note that the AbstractParserProvide provides a ByteVec and this lets you read as much data as you want in the guessSetup(...).

Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Wow, thanks. The whole chunkSize reevaluation is kind of clumsy there, and in static ParseSetup.guessSetup() there's no way to go ahead and get chunk size individually, per parser, but what we can do is use dataOffset that is already calculated in some parsers (ARFF), and go from there. if the parser does not calculate dataOffset, it's -1, so it has no impact. What do you think? Committing this update now.

**Comment by vpatryshev** [comment]
@h2o-ops please test!

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 4 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Sorry, it's not in master actually.
I'm going to use it everywhere, replacing the "width calculation" that is way too heavy and, as I have an example, faulty (have to investigate separately).

File: h2o-core/src/main/java/water/util/BytesStats.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Gathers stats for a byte array.
+ * 
+ * The array has to contain at least one NewLine character,
+ * otherwise it's considered bad, bad.
+ * 
+ * Created by vpatryshev on 3/27/17.
+ */
+public class BytesStats implements Serializable {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/BytesStats.java]: Serializables don't work with H2O's autobuffers. Should it be Iced instead?

File: h2o-core/src/test/java/water/parser/ChunksizeTest.java [hunk 9]
```diff
@@ -103,14 +104,18 @@ public void run() throws IOException {
...
+
+                  final double actual = totalSize / cloudSize / cores;
+                  Assert.assertTrue(
+                      "Total " + totalSize + ", cloud size" + cloudSize +", cores " + cores + " => " + actual, 
+                      actual > 1e9); //only for big data, where we have more than 1GB per core
                 }
 
                 if (fail) {
-                  Log.info(msg + (oldheuristic==0?"(New Heuristic)":"(Old Heuristic)"));
+//                  Log.info(msg + (oldheuristic==0?"(New Heuristic)":"(Old Heuristic)"));
```
Comment (michalkurka) [inline h2o-core/src/test/java/water/parser/ChunksizeTest.java]: Why is this commented out?

File: h2o-core/src/test/java/water/parser/ChunksizeTest.java [hunk 10]
```diff
@@ -122,17 +127,17 @@ public void run() throws IOException {
...
     fw.close();
       for (int i : new int[]{0,1}) {
-        Log.info((i==1 ? "Old" : "New") + " heuristic:");
-        Log.info("Total: " + counter[i]);
-        Log.info("Failure rate: " + PrettyPrint.formatPct((double) failed[i] / counter[i]));
-        Log.info("Too small: " + PrettyPrint.formatPct((double) toosmall[i] / counter[i]));
-        Log.info("Too large: " + PrettyPrint.formatPct((double) toolarge[i] / counter[i]));
-        Log.info("Too few: " + PrettyPrint.formatPct((double) toofew[i] / counter[i]));
-        Log.info("Too many: " + PrettyPrint.formatPct((double) toomany[i] / counter[i]));
+//        Log.info((i==1 ? "Old" : "New") + " heuristic:");
```
Comment (michalkurka) [inline h2o-core/src/test/java/water/parser/ChunksizeTest.java]: Why is it commented out?

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 11]
```diff
@@ -630,14 +598,15 @@ private static final long maxLineLength(byte[] bytes) {
...
+            maxLineLength = ll;
+          }
         }
-      } catch (IOException e) {
-        return -1;
-      }
+      } catch (IOException ignore) {}
       return maxLineLength;
     }
-    return -1;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Do you really not need to differentiate an empty file and input made of just empty lines? Why was the -1 there before?

File: h2o-core/src/main/java/water/util/BytesStats.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+  public final int numLines;
+  public final int maxWidth;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/util/BytesStats.java]: What does maxWidth mean? How does it handle international (non-ascii) characters?

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 13]
```diff
@@ -344,15 +352,15 @@ public GuessSetupTsk(ParseSetup userSetup) {
...
-        // only preview 1 DFLT_CHUNK_SIZE for ByteVecs, UploadFileVecs, compressed, and small files
-/*        if (ice instanceof ByteVec
-                || ((Frame)ice).vecs()[0] instanceof UploadFileVec
-                || bv.length() <= FileVec.DFLT_CHUNK_SIZE
-                || decompRatio > 1.0) { */
         try {
           _gblSetup = guessSetup(bv, bits, _userSetup);
+          _gblSetup.bytesStats = new BytesStats(bits);
+          {
+            if (_maxLineLength != _gblSetup.bytesStats.maxWidth) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This doesn't seem correct for non-ascii encodings.

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 14]
```diff
@@ -42,6 +45,9 @@
   String[][] _na_strings;       // Strings for NA in a given column
   String[][] _data;           // First few rows of parsed/tokenized data
 
+  int tentativeNumLines = -1;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This is only set and never used. What is tentativeNumLines for?

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 15]
```diff
@@ -42,6 +45,9 @@
...
+  int dataOffset = -1;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: My opinion is that tentativeNumLines & dataOffset should not be here. Two reasons: they only apply to the ARFF parser and they don't carry any meaningful information in the (unified) global setup (after two or more setups are merged in mergeSetups).

Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: I'll comment out tentativeNumLines, but dataOffset is needed for evaluating chunkSize, see line 281 of ParseSetup. I do not see any other way to fix chunkSize. Do you?
In any case, I'm planning to rewrite this stuff later on. But right now the value is required. It will be required in other parsers later on. You don't mean I should, while fixing this specific bug, go ahead and start changing other parsers? All in due time. Trust me.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 14 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This one is for investigating a stupid failure when line length is calculated to be 1.3M. Will get back to that case later on. Currently the line is commented out.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 13 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: Look at the next line, it's commented out. There's something fishy in data handling. This is the next thing to investigate. Let's keep it for a while. Once I add non-ascii data and reopen the case with wrong line length evaluation, it will do the right stuff.
Commenting it out for now.

File: h2o-core/src/main/java/water/util/BytesStats.java (hunk 12 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/util/BytesStats.java]: It means maximum number of bytes. E.g. for UTF8 it's the max number of bytes in UTF8 encoding. UTF8 seems to be our standard for internal strings storage. That's how data are stored currently in our BufferedStrings.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 11 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: An empty file contains exactly 0 bytes and 0 lines. That's the right answer. -1 is not the length of an empty file. You can try wc.
Craigs-MacBook-Pro-2:h2o-3 vpatryshev$ touch empty.txt
Craigs-MacBook-Pro-2:h2o-3 vpatryshev$ ls -l empty.txt
-rw-r--r--  1 vpatryshev  staff  0 Apr 11 14:26 empty.txt
Craigs-MacBook-Pro-2:h2o-3 vpatryshev$ wc empty.txt
       0       0       0 empty.txt
Craigs-MacBook-Pro-2:h2o-3 vpatryshev$

File: h2o-core/src/test/java/water/parser/ChunksizeTest.java (hunk 9 above)
Comment (vpatryshev) [inline h2o-core/src/test/java/water/parser/ChunksizeTest.java]: Because it gives no information, but fills the output with tons of unnecessary lines.

File: h2o-core/src/main/java/water/util/BytesStats.java (hunk 8 above)
Comment (vpatryshev) [inline h2o-core/src/main/java/water/util/BytesStats.java]: Ok, thanks, will try.

File: h2o-core/src/main/java/water/parser/ParseSetup.java [hunk 16]
```diff
@@ -270,8 +276,10 @@ public static ParseSetup guessSetup( Key[] fkeys, ParseSetup userSetup ) {
     if (ice instanceof Frame && ((Frame) ice).vec(0) instanceof UploadFileVec) {
       t._gblSetup._chunk_size = FileVec.DFLT_CHUNK_SIZE;
     } else {
-      t._gblSetup._chunk_size = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
-              Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int calculatedChunkSize = FileVec.calcOptimalChunkSize(t._totalParseSize, t._gblSetup._number_columns, t._maxLineLength,
+          Runtime.getRuntime().availableProcessors(), H2O.getCloudSize(), false /*use new heuristic*/, true);
+      int finalChunkSize = Math.max(t._gblSetup.dataOffset + 2, calculatedChunkSize);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: @vpatryshev I think we are close to a merge. I still don't understand how the dataOffset works for multiple files, there is no handling of dataOffset in the setup unification method (the one that creates the global offset).

Also I think it needs to be very clearly explained why is this being done here, otherwise this knowledge will get lost over time.

 Would it really not be possible to change ARFFParser/CsvParser to support multichunk headers? How hard would that be?

Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: This is a crutch, I understand.
Next I'm planning to slightly fix the parsers so they'll be able to deal with headers of any length, per file, or global, and not split input into chunks before carefully looking at it.
This current pr just solves one specific arrf problem.
I tried to move dataOffset to arrf, it's natural, but it did not work. The calculation of "chunk size" is happening where it is happening, and arrf parser has nothing to do with it, unfortunately. Changing it would mean redesigning the parsers. I hope to do it next.

File: h2o-core/src/test/java/water/parser/ChunksizeTest.java (hunk 10 above)
Comment (vpatryshev) [inline h2o-core/src/test/java/water/parser/ChunksizeTest.java]: See, this too has no informational value, and clutters the output greatly. It might have been good for testing, I don't know.

File: h2o-core/src/main/java/water/parser/ParseSetup.java (hunk 16 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: What will happen if we need to parse a directory that contains 2 ARFF files with different offsets?

Comment (vpatryshev) [inline h2o-core/src/main/java/water/parser/ParseSetup.java]: There is a sample test case for that.
The first one will define the chunk size. When the next one is opened, it will be too late to change it anyway. (I'm planning to fix this later on).
Generally speaking the impact of dataOffset is not so categorical; on many occasions the parsers will work even when the first dataOffset is 0. 
Good point anyway; I'll have to concoct a test case where the header is huge in the second file. It will probably fail with the current code. There are more cases that will fail, actually.

**Comment by abal5** [comment]
@h2o-ops please test

**Comment by abal5** [comment]
@h2o-ops please test

**Comment by abal5** [comment]
@h2o-ops please test

**Comment by abal5** [comment]
@h2o-ops please test

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by abal5** [comment]
@h2o-ops please test!
